src/Service/URLSanitisationService.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+namespace SilverStripe\StaticPublishQueue\Service;
+
+use SilverStripe\Core\Config\Configurable;
+use SilverStripe\Core\Environment;
+
+/**
+ * Class URLSanitisationService
+ * Provides an abstraction for populating and generating a list of
+ * URLs to be cached by StaticPublishQueue.
+ *
+ * @package SilverStripe\StaticPublishQueue\Service
+ */
+class URLSanitisationService
+{
+    use Configurable;
+
+    /**
+     * Enable to force all URLs to be changed to https
+     *
+     * @var bool
+     * @config
+     */
+    private static $force_ssl = false;
+
+    /**
+     * @var array
+     */
+    private $urls = [];
+
+    /**
+     * @var int
+     */
+    private $pointer = 0;
+
+    /**
+     * @param string $url
+     */
+    public function addURL(string $url): void
+    {
+        // We never want to have stage=Stage present in our cached URLs. This will safely remove "stage" params, but
+        // keep any others. It doesn't matter where in the string "stage=" exists.
+        $url = preg_replace('/([?&])stage=[^&]+(&|$)/', '$1', $url);
+        // Trim any trailing "?" or "&".
+        $url = $string = rtrim($url, '&');
+        $url = $string = rtrim($url, '?');
+
+        // Force SSL publishing as CLI will default to http
+        $forceSSL = $this->getForceSSL();
+        if ($forceSSL) {
+            $url = str_replace('http://', 'https://', $url);
+        }
+
+        // Check if this URL is already represented.
+        if (array_key_exists($url, $this->urls)) {
+            // Don't need to add it again if it is.
+            return;
+        }
+
+        $this->urls[$url] = $this->pointer;
+        $this->pointer += 1;
+    }
+
+    /**
+     * @param array $urls
+     */
+    public function addURLs(array $urls): void
+    {
+        foreach ($urls as $url) {
+            $this->addURL($url);
+        }
+    }
+
+    /**
+     * @return array
+     */
+    public function getURLs($formatted = true): array
```

Comment:
This is because the majority use of this function is to feed the URLs into the static cache job which requires the keyed format. I don't have a strong opinion on this though, so I'm happy to change the default.